feat: launcher gui

iamllama opened the pull request:
To run, install tauri's [system dependencies](https://v2.tauri.app/start/prerequisites/#system-dependencies), [cli (cargo)](https://v2.tauri.app/reference/cli) and then, in the repo root, `cargo tauri dev` with `.\yarn ld` in another window

- [ ] Error handling and display
  - [x] Check if os/glibc ver. is supported
- [x] Uninstall isn't implemented yet
- [ ] Refactors
  - [ ] The uv code was almost wholesale transplanted from the old launcher (with modifications for pty) and is in dire need of a cleanup but I didn't want to spend time on major changes without checking if any of this is acceptable (😅) and if we still want to keep the cli around?
  - [x] UI state needs reworking, versions should be queried on app launch, not on svelte mount
- [ ] CI
  - [ ] svelte-check action for the launcher
  - [ ] target for installing tauri cli tool
- [ ] UI design
Tentatively based it on the options pages (i probably won't be of much help here)

Closes #4152

<img width="300" alt="image" src="https://github.com/user-attachments/assets/e24a60e5-dbec-4d05-b81c-f8b9acad1cce" />


EDIT: CI is missing some of tauri's needed system deps

iamllama committed:
add crates

iamllama committed:
update package.json

iamllama committed:
codegen i18n ts for launcher

iamllama committed:
codegen launcher rpc (partly)

iamllama committed:
modify helper files to support launcher

iamllama committed:
add launcher strings

iamllama committed:
qt/launcher-gui

iamllama committed:
add to workspace, add helper scripts

iamllama committed:
modify i18n resources struct for launcher use

iamllama committed:
update minilints

iamllama committed:
update check_web (svelte-check still missing)

iamllama committed:
add option to hide revert buttons

iamllama committed:
add title slot

iamllama committed:
fix launcher scripts

iamllama committed:
update advisory allowlist...

dae commented:
This is impressive work llama! I'll need a bit more time to review the current code properly, but I wanted to give you some quick feedback to unblock you.

- Happy with the general direction, and don't think we'll want to keep around the terminal-based approach with this in place, subject to the next point. Given that, it might be nice if the code for this lived in qt/launcher, as that'll also make it easier to track any changes made to the old launcher code.
- We'll probably want to get an alpha build of this into the hands of testers as soon as basic functionality is working on all platforms, so we can identify any unanticipated compatibility issues that might prevent us from switching away from the terminal approach.
- Not urgent, but it would be nice if our build system eventually provides a target that takes care of installing the cli tool (like we do for cargo-nextest).
- Your pty approach saves us a bunch of work, and may end up being our best approach. Downside is that we won't be able to show overall progress to users who might be daunted by the per-package output. It might be worth starting a discussion with uv about possible alternatives, such as a flag that restricts progress to a single overall bar, or exposes progress events over a socket or the network. If they're amenable to something like that and we decide it's a good idea, we could send through a PR to make it easier for them to accept.
- I wasn't able to test the install process on a Mac, as I get the following during startup:

thread 'tokio-runtime-worker' panicked at qt/launcher-gui/src-tauri/src/commands.rs:140:58:
called `Result::unwrap()` on an `Err` value: Read "/Users/dae/Local/rust/Resources/.python-version"

The existing code expects the launcher binary to be run from within an app bundle. This is how the layout looks like for the terminal version:

```
$ ls -aR ../../../out/launcher/Anki.app
.        ..       Contents

../../../out/launcher/Anki.app/Contents:
.          ..         Info.plist MacOS      Resources

../../../out/launcher/Anki.app/Contents/MacOS:
.                 ..                install_name_tool launcher          uv

../../../out/launcher/Anki.app/Contents/Resources:
.               ..              .python-version Assets.car      pyproject.toml  versions.py
```

build.sh in qt/launcher/mac takes care of the building that folder after compiling. If you don't have access to a Mac, I can take care of this before the alpha test.

Should hopefully have more comments in the next few days!

iamllama committed:
add tauri-plugin-dialog crate

to make alert() work

iamllama committed:
add @xterm/addon-webgl package

xterm.js' new DOM renderer seems to have some visual glitches,
try using the webgl renderer first before falling back

iamllama committed:
add and use isWindows helper fn

iamllama committed:
add new rpc calls

iamllama committed:
update capabilities

iamllama committed:
refactor state and uv code

iamllama committed:
refactor app

iamllama committed:
remove unneeded rpc error message

iamllama committed:
refactor and implement rpc commands

iamllama committed:
change codegen to not alert by default on error

iamllama committed:
update tauri.conf.json

iamllama committed:
add strings

iamllama committed:
add font fallback

iamllama committed:
remove getVersions call

iamllama committed:
provide tr as store to avoid using {#key ...}

iamllama committed:
modify Warning to take icon

iamllama committed:
add CheckDecagramOutline icon

iamllama committed:
add components

iamllama committed:
refactor Start.svelte

iamllama committed:
add favicon

iamllama committed:
limit select menu height to 200px

iamllama committed:
make TitledContainer's title optional

iamllama committed:
./check

iamllama commented:
Pushed changes i made over the past week (sorry for the massive diffs!)

> Given that, it might be nice if the code for this lived in qt/launcher, as that'll also make it easier to track any changes made to the old launcher code.

I'd like to move it over after getting uninstall done, just to have it as a reference. The refactors i did will probably make tracking changes harder, as it included changing some of the fns to methods. But i'll likely revert that as it adds indentation for no gain

> It might be worth starting a discussion with uv about possible alternatives, such as a flag that restricts progress to a single overall bar, or exposes progress events over a socket or the network. If they're amenable to something like that and we decide it's a good idea, we could send through a PR to make it easier for them to accept.

I actually did look into this after encountering problems with pty on windows early on and was able to patch uv@0.7.13 to output parseable progress on stdout (pushed it here: https://github.com/astral-sh/uv/compare/0.7.13...iamllama:uv:sync-progress)

uv uses the indicatif crate to render those fancy progress bars on stderr, and if it isn't given a tty it doesn't show progress at all. I don't think uv uses stdout for anything so maybe we could use that instead of sockets/etc? It'd be best to @ zanieb to see what's acceptable (there's work being done on a JSON output format, but it doesn't include progress)

> I wasn't able to test the install process on a Mac, as I get the following during startup:

I don't have a mac, so help would be appreciated here 😅 I reckon a quick sanity check would be building the gui with `cargo tauri build`, modifying these paths in the old launcher's mac build script to point to launcher-gui instead:
https://github.com/ankitects/anki/blob/321e23acb2b02d657ca08727546c016d7ef27381/qt/launcher/mac/build.sh?plain=1#L28-L29
and then running the output of that script as per normal

iamllama committed:
update tr store in +page.svelte

iamllama committed:
remove unused stores

iamllama committed:
cleanup

iamllama committed:
remove snafu for now

iamllama committed:
add GetState rpc

iamllama committed:
add strings

iamllama committed:
add components

iamllama committed:
add uninstall button

iamllama committed:
extract normal state component out of Start

iamllama committed:
use getState

iamllama committed:
convert uv methods back to free fns

iamllama committed:
./ninja format

iamllama committed:
add todo to uninstall btn

iamllama committed:
add uninstall rpcs and state

iamllama committed:
add StateExt trait

iamllama committed:
add I18nExt trait

iamllama committed:
refactor to use traits

iamllama committed:
implement uninstall rpcs

iamllama committed:
update platform-specific uninstall fns

iamllama committed:
update handle_uninstall

iamllama committed:
unhardcode pty-data event name

iamllama committed:
add/modify strings

iamllama committed:
move font fallback to base.scss

iamllama committed:
fix spinner not being centred

iamllama committed:
implement uninstall screen

iamllama committed:
fetch and pass uninstall info

iamllama committed:
fix reactivity issue

this broke somewhere down the line and needs a tick now

iamllama committed:
cleanup

iamllama committed:
debug flag

iamllama committed:
remove use of setError

iamllama committed:
Merge branch 'main' into launcher-gui

iamllama committed:
port 739e41ce1

iamllama committed:
port a05c90cbc

iamllama committed:
port 23263caea

iamllama committed:
replace unwraps

iamllama committed:
modify strings

iamllama committed:
allow launching existing or installing new if either fails
